sky/data/data_utils.py

```diff
@@ -126,6 +146,175 @@ def verify_gcs_bucket(name: str) -> bool:
+
+    Args:
+      name: str; Name of AZ Container.
+
+    Returns:
+      boolean; shows either or not the container exists.
+    """
+    container_client = create_az_client('container', storage_account_name,
+                                        container_name)
+    return container_client.exists()
+
+
+def get_az_resource_group(
+        storage_account_name: str,
+        storage_client: Optional[Client] = None) -> Optional[str]:
+    """Returns the resource group name the given storage account belongs to.
+
+    Args:
+        storage_account_name: str; name of the storage account
+        storage_client: Optional[Client]; client object facing storage
+
+    Returns:
+        Name of the resource group the given storage account belongs to.
+    """
+    if storage_client is None:
+        storage_client = create_az_client('storage')
+    for account in storage_client.storage_accounts.list():
+        if account.name == storage_account_name:
+            # Extract the resource group name from the account ID
+            resource_group_name = account.id.split('/')[4]
+            return resource_group_name
+    # resource group cannot be found when using container not created
+    # under the user's subscription id, i.e. public container or
+    # private containers not belonging to the user.
+    return None
+
+
+def get_az_storage_account_key(
+    storage_account_name: str,
+    resource_group_name: Optional[str] = None,
+    storage_client: Optional[Client] = None,
+    resource_client: Optional[Client] = None,
+) -> Optional[str]:
+    """Returns access key of the given name of storage account.
+
+    Args:
+        storage_account_name: str; Name of the storage account
+        resource_group_name: str; Name of the resource group the passed storage
+            account belongs to.
+        storage_clent: Optional[Client]; client object facing Storage
+        resource_client: Optional[Client]; client object facing Resource
+
+    Returns:
+        One of the few access keys to the given storage account
+    """
+    # resource_group_name is None when using a public container or
+    # a private containers not belonging to the user.
+    if resource_group_name is None:
+        return None
+    if resource_client is None:
+        resource_client = create_az_client('resource')
+    if storage_client is None:
+        storage_client = create_az_client('storage')
+    resources = resource_client.resources.list_by_resource_group(
+        resource_group_name)
+    assert resources is not None
+    for resource in resources:
+        if (resource.type == 'Microsoft.Storage/storageAccounts' and
+                resource.name == storage_account_name):
+            storage_account_keys = storage_client.storage_accounts.list_keys(
+                resource_group_name, storage_account_name)
+            storage_account_keys = [
+                key.value for key in storage_account_keys.keys
+            ]
+    return storage_account_keys[0]
```

Comment:
The first time I ran `sky spot launch` with this branch, I got:
```
(base) ➜  romil_yamls git:(azure-blob-storage) ✗ sky spot launch task.yaml
Task from YAML spec: task.yaml
Traceback (most recent call last):
  File "/Users/romilb/tools/anaconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 354, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 805, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 354, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 805, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 3233, in spot_launch
    task_or_dag = _make_task_or_dag_from_entrypoint_with_overrides(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 758, in _make_task_or_dag_from_entrypoint_with_overrides
    dag = dag_utils.load_chain_dag_from_yaml(entrypoint, env_overrides=env)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/dag_utils.py", line 104, in load_chain_dag_from_yaml
    task = task_lib.Task.from_yaml_config(task_config, env_overrides)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/task.py", line 419, in from_yaml_config
    storage_obj = storage_lib.Storage.from_yaml_config(storage[1])
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1010, in from_yaml_config
    storage_obj.add_store(StoreType(store.upper()))
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 856, in add_store
    store = store_cls(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1925, in __init__
    super().__init__(name, source, region, is_sky_managed,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 243, in __init__
    self.initialize()
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2091, in initialize
    self.storage_account_key = data_utils.get_az_storage_account_key(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/data_utils.py", line 246, in get_az_storage_account_key
    return storage_account_keys[0]
UnboundLocalError: local variable 'storage_account_keys' referenced before assignment
```

The next time I ran, this message went away and I got:
```
(base) ➜  romil_yamls git:(azure-blob-storage) ✗ sky spot launch task.yaml
Task from YAML spec: task.yaml
<iterator object azure.core.paging.ItemPaged at 0x1273e5f40>
Traceback (most recent call last):
  File "/Users/romilb/tools/anaconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 354, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 805, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 354, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 805, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 3233, in spot_launch
    task_or_dag = _make_task_or_dag_from_entrypoint_with_overrides(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 758, in _make_task_or_dag_from_entrypoint_with_overrides
    dag = dag_utils.load_chain_dag_from_yaml(entrypoint, env_overrides=env)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/dag_utils.py", line 104, in load_chain_dag_from_yaml
    task = task_lib.Task.from_yaml_config(task_config, env_overrides)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/task.py", line 419, in from_yaml_config
    storage_obj = storage_lib.Storage.from_yaml_config(storage[1])
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1010, in from_yaml_config
    storage_obj.add_store(StoreType(store.upper()))
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 856, in add_store
    store = store_cls(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1925, in __init__
    super().__init__(name, source, region, is_sky_managed,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 243, in __init__
    self.initialize()
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2095, in initialize
    self.bucket_name, is_new_bucket = self._get_bucket()
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2246, in _get_bucket
    container = data_utils.create_az_client('container',
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/data_utils.py", line 166, in create_az_client
    return azure.get_client(client_type, subscription_id, storage_account_name,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/adaptors/common.py", line 61, in wrapper
    return func(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/adaptors/azure.py", line 91, in get_client
    container_client.exists()
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/storage/blob/_container_client.py", line 565, in exists
    process_storage_error(error)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/storage/blob/_shared/response_handlers.py", line 184, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
azure.core.exceptions.HttpResponseError: This request is not authorized to perform this operation using this permission.
RequestId:9ed80b39-501e-0086-6b03-9ce684000000
Time:2024-05-01T20:07:02.0302683Z
ErrorCode:AuthorizationPermissionMismatch
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>AuthorizationPermissionMismatch</Code><Message>This request is not authorized to perform this operation using this permission.
RequestId:9ed80b39-501e-0086-6b03-9ce684000000
Time:2024-05-01T20:07:02.0302683Z</Message></Error>
```

I was able to resolve this by manually adding Storage Blob Owner permission to the storage account created by SkyPilot, but we should perhaps add this permission when we create the storage account?

Reply:
1. `UnboundLocalError: local variable 'storage_account_keys' referenced before assignment` This error occurs as creation of storage account and resource group did not fully propagate to Azure yet, and it was resolved when I was testing it when I introduced this short term of `sleep` at [b67e87c](https://github.com/skypilot-org/skypilot/pull/3032/commits/b67e87c80dbab1d2c51bc571a43a7cad3cdbf705). I guess it can vary by the users. I'll add a try-catch block to note the users to try again. Should I increase the sleep time length as well?
2. Thanks for catching this. I realized not every account has the correct role by default when creating the storage account. Will update following your suggestion.

Reply:
The role assignment issue is resolved at [4c5938a](https://github.com/skypilot-org/skypilot/pull/3032/commits/4c5938ad4eece97a1d47c609eac3d6c08564dddc)